Implement copyright notice detection

Ryang20718 opened the pull request:
## Summary

Add copyright notice detection to enforce the presence of copyright headers in Python files.

Configurable settings include: the relevant regular expression, the author name, and the minimum file size, similar to [flake8-copyright](https://github.com/savoirfairelinux/flake8-copyright).

Closes https://github.com/charliermarsh/ruff/issues/3579

Ryang20718 commented on crates/ruff/src/checkers/physical_lines.rs:
```diff
@@ -154,6 +160,24 @@ pub(crate) fn check_physical_lines(
                 diagnostics.push(diagnostic);
             }
         }
+        if enforce_copyright_header && !checked_copyright_header_absent {
```
Could also do dep injection here and pass regexp as an arg. However, that would essentially put all the logic in physical_lines.rs. Let me know which you would prefer

Ryang20718 commented:
Hi, I made a first stab at implementing https://github.com/charliermarsh/ruff/pull/4701 flake8-copyright  

However, it seems like the "pre-commit" check in CI is failing due to a lack of copyright headers on some python files in the scripts directory https://github.com/charliermarsh/ruff/actions/runs/5115146921/jobs/9196121487?pr=4701
Wondering if it would be preferable to add the copyright header, disable the check or add noqa C801?

When I tried adding noqa, when emulating CI and running the command locally, it would overwrite my changes.

Ryang20718 commented:
Sorry if this was documented somewhere/asked again, but wondering if there's a group of people to assign as reviewers for PR's?

MichaReiser commented on crates/ruff/src/checkers/physical_lines.rs:
```diff
@@ -154,6 +160,24 @@ pub(crate) fn check_physical_lines(
+                }
+                None => (),
+            }
+            chars_before_copyright_header += line.len() as i64;
```
`line.len` gives you the number of bytes on that line, not the number of characters. The two numbers are different for non-ascii strings (e.g. emojis consist of multiple bytes, but represent a single character)

MichaReiser commented on crates/ruff/src/rules/flake8_copyright/rules/copyright_header_absent.rs:
```diff
@@ -0,0 +1,47 @@
+    settings: &Settings,
+    current_char_index: i64,
+) -> Option<bool> {
+    let copyright_regexp = format!(
+        "{} {}",
+        settings.flake8_copyright.copyright_regexp, settings.flake8_copyright.copyright_author
+    );
+    let regex = Regex::new(copyright_regexp.trim()).unwrap();
```
We should gracefully handle the case where the `Regex` is invalid rather than just panicking. 

@charliermarsh is there a place where we could do this deserialization outside of the rule (maybe in the configuration to settings deserialization?)

Ryang20718 replied:
Made it default to known string if user specified regex passed in for now, but open to better suggestions

MichaReiser commented on crates/ruff/src/rules/flake8_copyright/rules/copyright_header_absent.rs:
```diff
@@ -0,0 +1,47 @@
+    );
+    let regex = Regex::new(copyright_regexp.trim()).unwrap();
+
+    let out_of_range =
+        current_char_index > (settings.flake8_copyright.copyright_min_file_size as i64);
```
Ruff only supports files up to 4GB. Using a u32 should be fine.

Ryang20718 replied:
Will fix

MichaReiser commented on crates/ruff/src/rules/flake8_copyright/rules/copyright_header_absent.rs:
```diff
@@ -0,0 +1,47 @@
+    line: &Line,
+    settings: &Settings,
+    current_char_index: i64,
+) -> Option<bool> {
```
Nit: I think it would improve readability if this rule returns a custom enum instead of `Option<bool>`

```
#[derive(Copy, Clone, Eq, PartialEq)]
enum CopyrightHeaderKind{
	Missing,
	Present,
	TooSmall
}
```

Ryang20718 replied:
Will do!

MichaReiser commented on crates/ruff/src/rules/flake8_copyright/rules/copyright_header_absent.rs:
```diff
@@ -0,0 +1,47 @@
+/// format Copyright (C) <year> <author>
+///
+/// Error code C801
+pub(crate) fn copyright_header_absent(
```
Is it necessary that this rule runs on every line? Could we instead call the rule just once, passing the full source code? Should this rule only run on comments (or doc-strings)? rather than on all source code?

Ryang20718 replied:
![image](https://github.com/charliermarsh/ruff/assets/31294356/f2e95a9d-e4ea-47b5-a580-2f4104645a46)
I initially asked on discord. I think it would more efficient as well to pass the full source code as well.

Is there any rule that currently does this pattern or should I just create a separate portion in check_physical_lines to pass the entire source code?

MichaReiser replied:
@charliermarsh what's your take?

charliermarsh replied:
Yeah I'd suggest rewriting to call once per file, especially since it's limited to the first N characters anyway.

MichaReiser commented on crates/ruff/src/rules/flake8_copyright/settings.rs:
```diff
@@ -0,0 +1,83 @@
+pub struct Settings {
+    pub copyright_regexp: String,
+    pub copyright_author: String,
+    pub copyright_min_file_size: isize,
```
Does the rule use `-1` to indicate no limit? 

I think I would prefer the use of an `Option` instead. 


```suggestion
    pub copyright_min_file_size: Option<u32>,
```

Ryang20718 replied:
converted to u32, max will be 1024 chars with default always be. rule doesn't use -1

charliermarsh commented on crates/ruff/src/registry.rs:
```diff
@@ -736,6 +738,9 @@ pub enum Linter {
     #[prefix = "COM"]
     Flake8Commas,
     /// [flake8-comprehensions](https://pypi.org/project/flake8-comprehensions/)
+    #[prefix = "C80"]
```
Nit: let's use `CPY` here, I'd like to avoid overloading the `C` prefix.

charliermarsh commented on crates/ruff/src/rules/flake8_copyright/settings.rs:
```diff
@@ -0,0 +1,83 @@
+#[serde(
+    deny_unknown_fields,
+    rename_all = "kebab-case",
+    rename = "Flake8CopyrightOptions"
```
Can we do a brief GitHub Codesearch to see how often (if ever) these options are used? And if they're important to support?

Ryang20718 replied:
From a brief text based search on public repos,

Copyright min file size is used in 70+ places. https://github.com/search?q=copyright-min-file-size&type=code
Copyright regexp is used in 370+ places. https://github.com/search?q=copyright-regexp&type=code
Copyright Author is used in 290+ places https://github.com/search?q=%22copyright-author+%3D+%22&type=code

charliermarsh commented on crates/ruff/src/rules/flake8_copyright/settings.rs:
```diff
@@ -0,0 +1,83 @@
+// //! Settings for the `flake8-copyright` plugin.
```
Nit: double comment pragma here.

charliermarsh commented on crates/ruff/src/rules/flake8_copyright/rules/copyright_header_absent.rs:
```diff
@@ -0,0 +1,69 @@
+
+use crate::settings::Settings;
+
+use lazy_regex::Regex;
```
Nit: if we do need a lazy regex, lets just use `Lazy::new(|| Regex::new(...))`, to avoid adding another dependency.

Ryang20718 commented:
Addressed comments

This PR should be good for re-review. However, the CI problem still lingers. I tried adding `# noqa: CPY801` to the python files, but autofix removes the line. I could add a line to these scripts such as `# Copyright (c) 2023 ruff` or is there a way we can disable this check when running ruff on this repo for that particular github action?


![image](https://github.com/charliermarsh/ruff/assets/31294356/121b234e-88fd-45c1-9955-6a4d730eace6)

charliermarsh commented:
Sorry, missed that question -- can we add `CPY` to the `ignore` in `scripts/pyproject.toml`?

Ryang20718 commented:
TIL nested pyproject.toml. Neat!

Ryang20718 commented:
addressed all comments. Wondering if @charliermarsh or @MichaReiser could re-review please?

gertvdijk commented:
Hi! Cool to see a another new rule has landed.

Two comments - as seen in 0.0.237:

1.  The listing of linters appears inconsistent:
    ```
    $ ruff linter
    [...]
     COM flake8-commas
         CPY Copyright-related rules
      C4 flake8-comprehensions
    [...]
    ```
    That line should have been  `CPY flake8-copyright`, I think.

2. The docs say *This rule is part of the nursery, a collection of newer lints that are still under development. As such, it must be enabled by explicitly selecting CPY001.*, but enabled already using `CPY` (broader). Could it be an accidental copy/paste from some E-rules?

charliermarsh commented:
Thanks for this. I had intentionally renamed to "Copyright-related rules" but perhaps that was a mistake.

(I think it's probably a bug that `--select CPY` works... but it's not a big deal in this case since the category only has the one rule, so I might not even bother to fix :joy:)
